README.md

# usd - Ruby Class for SDM REST-API-Calls

There is a ruby class and a commandline tool rusdc.

[![asciicast](https://asciinema.org/a/7zw3RLpikFluqX9XJMxCpMEmS.svg)](https://asciinema.org/a/7zw3RLpikFluqX9XJMxCpMEmS)

# installation

just install the gem:

```bash
gem install usd
```

Be aware that you need some compiler-tools and the ruby headers:

```bash
# ubuntu
sudo apt-get install build-essential ruby-dev
# redhat
yum group install "Development Tools"
yum install ruby-devel
```

# documentation

- see [https://github.com/ogaida/usd/wiki](https://github.com/ogaida/usd/wiki)
- if you are looking for dependencies in between the objects, objectnames or relational objects, then have a look at [techdocs.broadcom.com](https://techdocs.broadcom.com/content/broadcom/techdocs/us/en/ca-enterprise-software/business-management/ca-service-management/14-1/reference/ca-service-desk-manager-reference-commands/objects-and-attributes.html)

## bash-completion

see [https://github.com/ogaida/usd/wiki/bash-completion](https://github.com/ogaida/usd/wiki/bash-completion)

# external tools / requirements

If you use the `rusdc find` command with `--format mlr` option then you need `mlr`.

- [mlr - Miller](http://johnkerl.org/miller/doc/index.html) - a great tool for data-transforming to and from json, csv and many more

Place the `mlr`-binary in a path, which is in your PATH-Environment. Download-Url for mlr-releases: [https://github.com/johnkerl/miller/releases](https://github.com/johnkerl/miller/releases).

# functions from the commandline-tool `rusdc`

```
rusdc chg_add_attachment <co> <file>                                         # attach the file to co
rusdc chg_add_nr <co> <ci>                                                   # add a CI to a changeorder
rusdc chg_list_nr <co>                                                       # list all CIs of one changeorder
rusdc create                                                                 # pipe json-data to create object
rusdc field_names <object-type> [wc]                                         # list all fields of an object including its format
rusdc find <object-type> [where-clause] [fields, comma separated] [options]  # finds each object which meets the wc-condition
rusdc get <object-type> <common_name|id>                                     # shows one object by name or id
rusdc get_all_attachments_of_co <co_name>                                    # get all attachments of a changeorder and save all these to current folder
rusdc get_attachment_of_ci <ci_name> <filename>                              # download an attachment of a CI and print it out on stdout
rusdc get_attachment_of_co <co_name> <filename>                              # download an attachment of a changeorder and print it out on stdout
rusdc get_attachment_of_in <incident> <filename>                             # download an attachment of an Incident and print it out on stdout
rusdc help [COMMAND]                                                         # Describe available commands or one specific command
rusdc in_add_2_chg <changeorder> <incident>                                  # add incident to changeorder
rusdc in_add_attachment <in> <file>                                          # attach a file to an incident or problem
rusdc in_list_child_ins <in>                                                 # list all child-incidents of one incident
rusdc list_attachments_of_ci <ci_name>                                       # list all attachments of a CI
rusdc list_attachments_of_co <co_name>                                       # list all attachments of a changeorder
rusdc list_attachments_of_in <in>                                            # list all attachments of an incident or problem
rusdc nr_add_attachment <nr> <file>                                          # attach the file to nr
rusdc nr_add_child <nr-name> <child-name>                                    # add one child CI to another CI
rusdc nr_add_contact <nr> <cnt_email>                                        # add contact to configuration-item
rusdc nr_add_org <nr> <org>                                                  # add an Organisation to a ConfigurationItem
rusdc nr_changes <nr> [inactive-too]                                         # list all open changeorders of one CI
rusdc nr_childs <ci-name>                                                    # lists all childs CIs of a specific CI
rusdc nr_incidents <nr> [inactive-too]                                       # lists all incident of a specific CI
rusdc nr_parents <ci-name>                                                   # lists all parent CIs of a specific CI
rusdc update                                                                 # pipe json-data to update object
rusdc update_attr <obj> <common_name|wc> <key> <value>                       # updates a direct (not referenced) attribute of one or more objects.
rusdc update_attr_by_id <obj> <id> <key> <value>                             # updates a plain attribute of one object by id, does not work
rusdc update_ref_attr <obj> <common_name> <key> <value> [ref_obj=nr]         # updates a referenced attribute of one object.
rusdc update_ref_attr_by_id <obj> <id> <key> <value_id>                      # updates a referenced attribute of one object_id by value-id
rusdc vcard_via_email <email>                                                # creates a vcard from the given email-address and saves it in the /tmp directory.
```

# environment-variables

rusdc needs some environment-variables. there are:

```
usduser
usdpass
usdurl
```

you may set and export them via script:

```bash
#!/usr/bin/env bash
# load with `source ./set_env`

export usduser=<your username>
export usdurl="http://<your sdm server>:8050"
if [ -z "$usdpass" ]
then
  read -s -p "password for $usduser ? " usdpass
  echo ""
  export usdpass
fi
```

# example - find records

```bash
$ rusdc find nr "name like 'foo%'" name,description --format mlr
name  description
foo1  foo QS
foo2  foo TEST
```

more examples under [find examples-Wikipage](https://github.com/ogaida/usd/wiki/find-examples) or in the [change_log](https://github.com/ogaida/usd/blob/master/change_log.md)

# senarios

<img src="https://yuml.me/diagram/scruffy/class/[SDM%20REST-API{bg:red}]<->[usd-gem{bg:green}]<->[rusdc{bg:orange}],[ruby-script{bg:orange}]<->[usd-gem],[sinatra%20web-app{bg:orange}]<->[usd-gem],[bash commands]<->[rusdc],[shell-scripts]<->[rusdc],[import]->[ruby-script],[ruby-script]->[export],[reports]<-[sinatra%20web-app]<-[drop datafiles/emails/create objects]">

# feedback

Since I could only test this gem in a 14.1 environment, I would be interested to know whether it also works in a version 17 environment of CA Service Management. Any kind of feedback is very welcome.


change_log.md
# rusdc changelog

## 1.5.2

- disabled output, when login was successful. before this, it led to errors when the output was parsed

## 1.5.0

- sdm-access token caching changed, now for multiple user-sessions in one environment

## 1.4.2

- count number changed for search method from fixed value of 50 to dynamic value

## 1.4.1

- changed some comments

## 1.4.0

- remove "warning: URI.escape is obsolete"

## 1.3.7

- Bugfix in rusdc-function `get_attachment_of_co`

## 1.3.6

- Bugfix : uploaded filename with special character will be renamed with `URI.escape`

## 1.3.5

- new function `nr_add_contact <nr> <email>` to assign a contact to a configuration-item

## 1.3.4.1

- fixed another error in Usd.search. which comes with 1.3.1 .

## 1.3.4

- move function `upload_attachment` from `bin/rusdc` to `lib/usd`.

## 1.3.3

- private `rusdc`-funktion `upload_attachment` changed. Instead using `base64` and `unix2dos` binaries, we use Base64-Ruby-Class now. It is already required by usd.

## 1.3.2

- fixed error in Usd.search. which comes with 1.3.1 .

## 1.3.1

- new functions:
  - `chg_add_attachment`
  - `nr_add_attachment`
  - `in_add_attachment`
  - `vcard_via_email`

with the first three funtions you are able to upload files to changeorders, configurationitems and incidents and it does not matter if it is text or binary. the last new function creates vcards in version 2.1 from a contact via its email as identifier.

## 1.2

- now in Usd.search function:
  - whereclause is enriched with "common_name like '% wc%'" if it is not a valid whereclause.

## 1.0 and 1.1

- `update_ref_attr` rusdc-function fetches the IDs to update the object by itself.

## 0.3

- some common_name resolution

```ruby
class Usd
  CN={
    "chg" => "chg_ref_num",
    "cnt" => "combo_name",
    "arcpur_rule" => "name",
    "ca_tou" => "name",
    "cost_cntr" => "name",
    "country" => "name",
    "dept" => "name",
    "gl_code" => "name",
    "job_func" => "name",
    "loc" => "name",
    "nr" => "name",
    "opsys" => "name",
    "org" => "name",
    "tab" => "name",
    "auto_close" => "sym",
    "aty" => "sym",
    "act_type_assoc" => "sym",
    "ca_cmpny" => "sym",
    "closure_code" => "sym",
    "cmth" => "sym",
    "symptom_code" => "sym",
    "state" => "sym",
    "crt" => "sym",
    "ctab" => "sym",
    "ctp" => "sym",
    "dcon_typ" => "sym",
    "dlgsrvr" => "sym",
    "dmn" => "sym",
    "doc_rep" => "sym",
    "fmgrp" => "sym",
    "ical_alarm" => "sym",
    "ical_event_template" => "sym",
    "imp" => "sym",
    "intfc" => "sym",
    "kwrd" => "sym",
    "mfrmod" => "sym",
    "mgsstat" => "sym",
    "nrf" => "sym",
    "options" => "sym",
    "outage_type" => "sym",
    "perscnt" => "sym",
    "position" => "sym",
    "pr_trans" => "sym",
    "prod" => "sym",
    "quick_tpl_types" => "sym",
    "rc" => "sym",
    "resocode" => "sym",
    "resomethod" => "sym",
    "response" => "sym",
    "rrf" => "sym",
    "rss" => "sym",
    "seq" => "sym",
    "sev" => "sym",
    "site" => "sym",
    "slatpl" => "sym",
    "special_handling" => "sym",
    "svc_contract" => "sym",
    "typecnt" => "sym",
    "tz" => "sym",
    "tspan" => "sym",
    "transition_type" => "sym",
    "urg" => "sym",
    "vpt" => "sym",
    "wrkshft" => "sym",
    "cr" => "ref_num",
    "in" => "ref_num",
    "pr" => "ref_num"
  }
end
```

## 0.2.5.3

- neu function : `rusdc nr_add_org <nr> <org>  # add an Organisation to a ConfigurationItem`

#### fixes:

- rusdc - line 110: more robust when querying date fields

## 0.2.4.2

the same fixes as in 0.2.4.1 for:

```
rusdc list_attachments_of_ci <ci_name>
rusdc list_attachments_of_co <co_name>    
```

## 0.2.4.1

### fixes

- `rusdc get_attachment_of_ci` : attachment with spaces in path could not be downloaded

## 0.2.4

- find function has exitcode > 0 if no records has been found

```
$ rusdc find nr "name = 'not_here'"  || echo 'i am not here, please create me'
[]
i am not here, please create me
$ rusdc find cnt "last_name = 'Gaida'"  && echo 'i am here, you may update me'
[{
  "@COMMON_NAME": "Gaida, Oliver "
}]
i am here, you may update me
```

### fixes

- `rusdc get_all_attachments_of_co` : attachment with spaces in path could not be downloaded

## 0.2.3

- new function `in_add_2_chg` - add incident to changeorder

Usage:

```

  rusdc in_add_2_chg <changeorder> <incident>
```

Example:

```
$ rusdc find in "change.chg_ref_num = 'CO000001'"  --format mlr
@COMMON_NAME
I000001
I000002
$ rusdc in_add_2_chg CO000001 I000003
$ rusdc find in "change.chg_ref_num = 'CO000001'"  --format mlr
@COMMON_NAME
I000001
I000002
I000003
```

## 0.2.2

- extented function `update_attr`, now you may use a where-clause instead of a `common_name` to filter the object(s) to update

Example 1:

update many CIs with one statement...b

```bash
$ rusdc find nr "name like 'server[12]'" name,description --format mlr
name    description
server1 vorher
server2 vorher
$ rusdc update_attr nr "name like 'server[12]'" description "new description"
$ rusdc find nr "name like 'server[12]'" name,description --format mlr
name    description
server1 new description
server2 new description
```

Example 2:

update a CI which has a special attribute value:

```bash
$ rusdc find nr "serial_number = 'ABC123ABC987'" name,serial_number,warranty_end --format mlr
name    serial_number warranty_end
server1 ABC123ABC987  2020-11-30 00:00:00 +0100
$ rusdc update_attr nr "serial_number = 'ABC123ABC987'" warranty_end "01.02.2022"
$ rusdc find nr "serial_number = 'ABC123ABC987'" name,serial_number,warranty_end --format mlr
name    serial_number warranty_end
server1 ABC123ABC987  2022-02-01 00:00:00 +0100
```

## 0.2.1

- function `update_attr` has a new option for putting plain text with linebreaks in it.

- new funciton `update_ref_attr_by_id` sometimes there are problem if you try to update via common_name, so here you can use the `@id` instead.

- `rusdc get` now supports id too


lib/usd.rb
#

require 'rest-client'
require "base64"
require "json"
require "yaml"
require 'uri'
require "time"
require 'tempfile'

class Usd

  CN={
    "grc" => "type",
    "chg" => "chg_ref_num",
    "cnt" => "combo_name",
    "arcpur_rule" => "name",
    "ca_tou" => "name",
    "cost_cntr" => "name",
    "country" => "name",
    "dept" => "name",
    "gl_code" => "name",
    "job_func" => "name",
    "loc" => "name",
    "nr" => "name",
    "opsys" => "name",
    "org" => "name",
    "tab" => "name",
    "auto_close" => "sym",
    "aty" => "sym",
    "act_type_assoc" => "sym",
    "ca_cmpny" => "sym",
    "closure_code" => "sym",
    "cmth" => "sym",
    "symptom_code" => "sym",
    "state" => "sym",
    "crt" => "sym",
    "ctab" => "sym",
    "ctp" => "sym",
    "dcon_typ" => "sym",
    "dlgsrvr" => "sym",
    "dmn" => "sym",
    "doc_rep" => "sym",
    "fmgrp" => "sym",
    "ical_alarm" => "sym",
    "ical_event_template" => "sym",
    "imp" => "sym",
    "intfc" => "sym",
    "kwrd" => "sym",
    "mfrmod" => "sym",
    "mgsstat" => "sym",
    "nrf" => "sym",
    "options" => "sym",
    "outage_type" => "sym",
    "perscnt" => "sym",
    "position" => "sym",
    "pr_trans" => "sym",
    "prod" => "sym",
    "quick_tpl_types" => "sym",
    "rc" => "sym",
    "resocode" => "sym",
    "resomethod" => "sym",
    "response" => "sym",
    "rrf" => "sym",
    "rss" => "sym",
    "seq" => "sym",
    "sev" => "sym",
    "site" => "sym",
    "slatpl" => "sym",
    "special_handling" => "sym",
    "svc_contract" => "sym",
    "typecnt" => "sym",
    "tz" => "sym",
    "tspan" => "sym",
    "transition_type" => "sym",
    "urg" => "sym",
    "vpt" => "sym",
    "wrkshft" => "sym",
    "cr" => "ref_num",
    "in" => "ref_num",
    "pr" => "ref_num"
  }

  RestClient.proxy = ""

  attr_reader :base_url, :user, :access_key, :expiration_date, :debug
  attr_writer :debug

  def initialize(user, password="", base_url = "http://localhost:8050", hash={})
    hash = {:expiration_date => 0, :access_key => "",:save_access_key => true }.update hash
    make_new = true
    @base_url = base_url
    @user = user
    @debug = false
    remfile = "#{ENV["HOME"]}/.usd_#{user}"
    if File.exist?(remfile) and hash[:save_access_key]
      tt = YAML.load(File.open(remfile,"r"))
      if (tt.expiration_date - Time.now.to_i ) > 900 and tt.base_url == base_url
        @access_key = tt.access_key
        @expiration_date = tt.expiration_date
        make_new = false
        puts "cache key loaded..." if @debug
      end
    end
    if hash[:save_access_key] and hash[:access_key].length > 0
      if (hash[:expiration_date] - Time.now.to_i ) > 900
        @access_key = hash[:access_key]
        @expiration_date = hash[:expiration_date]
        make_new = false
        puts "use existing access_key ..." if @debug
      end
    end
    if make_new
      encoded=Base64.encode64( "#{@user}:#{password}")
      begin
        response = RestClient::Request.execute(method: :post, url: "#{@base_url}/caisd-rest/rest_access?_type=json",
          payload: '<rest_access/>',
          headers: {
            'content-type' => "application/xml",
            "accept" => "application/json",
            "authorization" =>  "Basic #{encoded}",
            "cache-control" => "no-cache"
          },
          log: Logger.new("/dev/null")
        )
        authData=JSON.parse(response.body)
        if authData['rest_access']['access_key'] > 0
          @access_key = authData['rest_access']['access_key']
          @expiration_date = authData['rest_access']['expiration_date']
          if hash[:save_access_key]
            f=File.open(remfile,"w")
            f.puts self.to_yaml
            f.close
          end
        else
          "keinen Accesskey erhalten. \nresponse.body:\n#{response.body}"
        end
      rescue RestClient::ExceptionWithResponse => e
        e.response
      end
    end
  end

  def self.loadcon
    # load from env
    Usd.new(ENV["usduser"],ENV["usdpass"],ENV["usdurl"])
  end

  def header(hash={})
    hash = {
      'x-accesskey' => @access_key,
      'accept' => "application/json",
      "Content-Type" =>"application/json; charset=UTF-8",
      'X-Obj-Attrs' => "*",
      'cache-control'=> "no-cache"
    }.update hash
    hash
  end

  def request(uri, hash={})
    RestClient.log = STDOUT if @debug
    hash = {:method => "get", :header => header(), :unchanged => false, :json => "", :base_url => @base_url}.update hash
    puts "request - hash: #{JSON.pretty_generate(hash)}" if @debug
    parser = URI::RFC2396_Parser.new
    if (uri !~ /^http/)
      url = parser.escape("#{hash[:base_url]}#{uri}")
    else
      url = parser.escape(uri)
    end
    begin
      if hash[:method] == "get"
        response = RestClient::Request.execute(method: hash[:method], url: url, headers: hash[:header])
      elsif   hash[:method] == "post"
        response = RestClient.post(url, hash[:json], hash[:header])
      elsif   hash[:method] == "put"
        response = RestClient.put(url, hash[:json], hash[:header])
      elsif   hash[:method] =~ /delete/i
        response = RestClient.delete(url, hash[:header])
      end
      if hash[:unchanged]
        response.body
      else
        JSON.parse(response.body)
      end
    rescue RestClient::ExceptionWithResponse => e
      e
    end
  end

  def create(hash = {})
    hash = {:type => "ruby", :data => {}}.update hash
    case hash[:type]
    when "ruby"
      data = hash[:data]
    when "json"
      data = JSON.parse(hash[:data])
    when "yaml"
      data = YAML.load(hash[:data])
    else
      "Error: 'data[:type]': '#{hash[:data]}' not found!"
    end
    puts "create - data: #{JSON.pretty_generate(data)}" if @debug
    object = data.keys[0]
    uri = "/caisd-rest/#{object}"
    request("/caisd-rest/#{object}",{:method => "post", :json => data.to_json})
  end

  def update(hash = {})
    hash = {:type => "ruby", :data => {}}.update hash
    case hash[:type]
    when "ruby"
      data = hash[:data]
    when "json"
      data = JSON.parse(hash[:data])
    when "yaml"
      data = YAML.load(hash[:data])
    else
      "Error: 'data[:type]': '#{hash[:data]}' not found!"
    end
    puts "update - data: #{JSON.pretty_generate(data)}" if @debug
    object = data.keys[0]
    if  data[object].has_key?("@id")
      request("/caisd-rest/#{object}/#{data[object]["@id"]}",{:method => "put", :json => data.to_json, :header => header({'X-Obj-Attrs' => 'COMMON_NAME'})})
    elsif data[object].has_key?("@COMMON_NAME")
      request("/caisd-rest/#{object}/COMMON_NAME-#{data[object]["@COMMON_NAME"]}",{:method => "put", :json => data.to_json, :header => header({'X-Obj-Attrs' => 'COMMON_NAME'})})
    else
      puts "specify @COMMON_NAME or @id at least."
    end
  end

  def set_url_parm(params_hash,attribute_name,default)
    # disable case
    params_hash.keys.each do |k|
      if k =~ /^#{attribute_name}$/i
        v = params_hash[k]
        params_hash.delete(k)
        params_hash[attribute_name] = v
      end
    end
    "#{attribute_name}=#{params_hash[attribute_name]?params_hash[attribute_name]:default}"
  end

  def set_param(params_hash,attribute_name,default)
    params_hash[attribute_name]?params_hash[attribute_name]:default
  end

  def search(object,params={})
    attr=[]
    attr.push set_url_parm(params,"SORT","id DESC")
    attr.push set_url_parm(params,"start","1")
    attr.push set_url_parm(params,"size","50")
    attr.push set_url_parm(params,"WC","")
    # if wc contains no sql-compare operater, it will be changed to "COMMON_NAME like '%<wc-before>%'"
    wc = attr.pop
    wc.gsub!(/^WC=/,'')
    if ([" like ","<",">","="," is "," in "," LIKE ", " IS ", " IN "].find {|e| wc =~ /#{e}/}).nil?
      wc = "#{CN[object]} like '%#{wc}%'"
    end
    attr.push "WC=#{wc}"
    # puts attr.jp # debug
    fields = set_param(params,"fields","COMMON_NAME,id")
    query_string=attr.join("&")
    res_rdata = request("/caisd-rest/#{object}?#{query_string}",{:method => "get", :header => header({'X-Obj-Attrs' => fields})})
    puts res_rdata if @debug
    count = res_rdata["collection_#{object}"]["@COUNT"].to_i
    start = res_rdata["collection_#{object}"]["@START"].to_i
    total_count = res_rdata["collection_#{object}"]["@TOTAL_COUNT"].to_i
    # turn throught the pages
    if count == 1
      [res_rdata["collection_#{object}"][object]]
    elsif count == 0
      []
    else
      retArray = res_rdata["collection_#{object}"][object]
      if total_count > (count + start - 1)
        new_params = {"start" => (start + count)}
        params = params.update new_params
        retArray += search(object,params)
      end
      retArray
    end
  end

  def upload_attachment(file, baseurl = @base_url, ak = @access_key)
    baseurl =~ /^([^:]+):/
    server = $1
    filename = File.basename(file)
    #filename_escaped = URI.escape(filename)
    filename_escaped = filename.gsub(/[^0-9A-Za-z.\-]/, '_')
    uri = "/caisd-rest/attmnt?repositoryId=1002&serverName=#{server}&mimeType=Text&description=#{filename_escaped}"
    url = URI("#{baseurl}#{uri}")
    http = Net::HTTP.new(url.host, url.port);
    request = Net::HTTP::Post.new(url)
    request["X-AccessKey"] = ak
    request["Content-Type"] = "multipart/form-data; BOUNDARY=*****MessageBoundary*****"
    request["accept"] = "application/json"
    request["Cache-Control"] = "no-cache"
    fileObj = File.open(file, "rb")
    fileContent = fileObj.read
    fileObj.close
    request.body = "--*****MessageBoundary*****\r\n \r\nContent-Disposition: form-data; name=\"payload\" \r\nContent-Type: application/xml; CHARACTERSET=UTF-8 \r\n\r\n \r\n<attmnt> \r\n<repository id=\"1002\"></repository> \r\n<orig_file_name>#{filename_escaped}</orig_file_name> \r\n<attmnt_name>#{filename_escaped}</attmnt_name> \r\n<description>Uploaded with rusdc from rubygem usd</description> \r\n</attmnt> \r\n\r\n \r\n--*****MessageBoundary*****\r\n \r\nContent-Disposition: form-data; name=\"#{filename_escaped}\"; filename=\"#{filename_escaped}\" \r\nContent-Type: application/octet-stream \r\nContent-Transfer-Encoding: base64\r\n\r\n#{Base64.encode64(fileContent)}\r\n\r\n \r\n--*****MessageBoundary*****--\r\n"
    response = http.request(request)
    response.read_body
  end

end

module Jsonpretty
  def jp
    # return in json_pretty
    JSON.pretty_generate(self)
  end
  def jpp
    # print in json_pretty
    puts JSON.pretty_generate(self)
  end
end

class Hash
  include Jsonpretty
end

class Array
  include Jsonpretty
end
